src/sentry/api/endpoints/user_emails.py

```diff
@@ -0,0 +1,240 @@
+
+
+def get_email(request):
+    """
+    Returns None if invalid email else returns a normalized email (lowercase + whitespace strip)
+    """
+
+    serializer = EmailSerializer(data=request.DATA)
+
+    # Bad email
+    if not serializer.is_valid():
+        return None
+
+    return serializer.object['email'].lower().strip()
+
+
+def add_email(email, user):
+    """
+    Adds an email to user account
+
+    Can be either primary or secondary
+    """
+
+    # Bad email
+    if email is None:
+        raise InvalidEmailError
+
+    # check if this email already exists for user
+    if email and UserEmail.objects.filter(
+        user=user, email__iexact=email
+    ).exists():
+        raise DuplicateEmailError
+
+    try:
+        with transaction.atomic():
+            new_email = UserEmail.objects.create(user=user, email=email)
+    except IntegrityError:
+        raise DuplicateEmailError
+    else:
+        new_email.set_hash()
+        new_email.save()
+        user.send_confirm_email_singular(new_email)
+
+        # Update newsletter subscription and mark as unverified
+        newsletter.update_subscription(user=user,
+                                       verified=False,
+                                       )
+        return new_email
+
+
+class UserEmailsEndpoint(UserEndpoint):
+    def get(self, request, user):
+        """
+        Get list of emails
+        ``````````````````
+
+        Returns a list of emails. Primary email will have `isPrimary: true`
+
+        :auth required:
+        """
+
+        emails = user.emails.all()
+
+        return Response(serialize(list(emails), user=user))
+
+    @sudo_required
+    def post(self, request, user):
+        """
+        Adds a secondary email address
+        ``````````````````````````````
+
+        Adds a secondary email address to account.
+
+        :param string email: email to add
+        :auth required:
+        """
+
+        email = get_email(request)
+
+        try:
+            new_email = add_email(email, user)
+        except (InvalidEmailError, DuplicateEmailError):
+            return InvalidEmailResponse
+        else:
+            logger.info(
+                'user.email.add',
+                extra={
+                    'user_id': user.id,
+                    'ip_address': request.META['REMOTE_ADDR'],
+                    'email': new_email.email,
+                }
+            )
+
+            return Response(status=status.HTTP_204_NO_CONTENT)
+
+    @sudo_required
+    def put(self, request, user):
+        """
+        Updates primary email
+        `````````````````````
+
+        Changes primary email
+
+        :param string email: the email to set as primary email
+        :auth required:
+        """
+
+        new_email = get_email(request)
+        old_email = user.email
+
+        if new_email is None:
+            return InvalidEmailResponse
+
+        # If email doesn't exist for user, attempt to add new email
```

Comment:
don't we prevent people from changing their primary email address until it's verfied?

Reply:
Hmm, it looks like on the "Emails" tab it hides the UI button to set a primary email until it's verified. However, in "Account Details", you can change the email address.